### DISCUSSION TITLE
Custom polyfills

### DISCUSSION THREAD

**Discussion by Nick-Mazuk** [vercel/next.js#20992] [Help] [answered]
In the [NextJS docs on polyfills](https://nextjs.org/docs/basic-features/supported-browsers-features), it says to add custom polyfills via `_app.tsx`. Is there an example for this?

Specifically, I'm looking for a way to load custom polyfills only for browsers that need them. For instance, a ResizeObserver or IntersectionObserver polyfill.

P.S. The [example in the examples folder](https://nextjs.org/docs/basic-features/supported-browsers-features) says that that's not the recommended way of loading polyfills. It just references the docs for custom polyfills.

**Comment by leerob** [comment, marked as answer]
Good catch — this isn't very clear right now. You should add a top-level import for the specific polyfill you need in your Custom `<App>` or the individual component. Example:

```js
import 'resize-observer-polyfill'

function MyApp({ Component, pageProps }) {
  return <Component {...pageProps} />
}

export default MyApp
```

I will update the example 👍

**Reply by leerob** [reply]
https://github.com/vercel/next.js/pull/20996

**Reply by Nick-Mazuk** [reply]
Great! And then NextJS will only serve the polyfill to browsers that need it? If so, that's awesome.

**Reply by wereHamster** [reply]
@Nick-Mazuk no, the polyfill will always be included, for all browsers.

**Reply by wereHamster** [reply]
Also note that if you import polyfills in _app, and have have a code editor plugin which automatically sorts imports, you may end up in a situation that the polyfill is loaded too late, after code that needs it. Therefore I'm not a fan of this solution.

What I use is put all imports into a separate file (src/polyfills.ts), and then make sure it's loaded as very first code through custom webpack config:

```
  webpack: (config) => {
    const originalEntry = config.entry
    config.entry = async () => {
      const entries = await originalEntry()

      if (entries['main.js']) {
        if (!entries['main.js'].includes('./src/polyfills.ts')) {
          entries['main.js'].unshift('./src/polyfills.ts')
        }
      }

      return entries
    }

    return config
  }
```

**Reply by leerob** [reply]
> Also note that if you import polyfills in _app, and have have a code editor plugin which automatically sorts imports, you may end up in a situation that the polyfill is loaded too late, after code that needs it. Therefore I'm not a fan of this solution.

I disagree. A single line import is much easier than overriding the webpack config. You can just add a lint ignore to that line?

**Reply by wereHamster** [reply]
But you have to remember to do that. The danger is that you forget and suddenly something doesn't work and you don't know why. Especially since the example makes no mention of that danger.

I found another approach to add polyfills that doesn't require any changes to the webpack config (but also works only for certain types of polyfills): instead of importing them in _app, load them via a <script> tag from a local api route, using the same logic as polyfill.io to deliver just the code that is required by the UA. For example, Safari 13 doesn't support Intl.RelativeTimeFormat, and polyfilling that requires ~150k JS (minified). It would be nice to avoid loading that in browsers which support that feature (same with ResizeObserver, IntersectionObserver etc). This approach works only for code which is required in the UA, as the polyfills are only applied in the client, but not on the server. This approach avoids relying on third-party services, which isn't advantageous anyways now that all UAs partition cache by origin. See example here: https://codesandbox.io/s/nextjs-api-polyfill-y9gii

**Reply by paul-vd** [reply]
Thanks @wereHamster for that solution.

I was busing doing a demo for clients and to my surprise when I wanted to show them the new feature I could not understand why my app was returning that a function is not defined on a prototype that I just presently created a custom polyfill for. going back to my reverted code I see that when I applied my import sorter via vs code, it set the polyfill after the required code.

Whilst I agree with you @leerob about the simplicity of importing in the _app, currently there is not "ignore" comments you can add to your file to avoid vscode from organizing your imports, which can always lead to human error of forgetting that the polyfill should be first.

![image](https://user-images.githubusercontent.com/39598117/110488920-39313f80-80ef-11eb-8631-27b6ac1c4b4f.png)
https://github.com/microsoft/TypeScript/issues/41494

So in a production environment it is much safer to do the former solution and just customize the webpack.

**Reply by tettoffensive** [reply]
@leerob This doesn't seem to work with 'css-paint-polyfill'. It complains that `window` is not defined. Perhaps there's a way to do this client-side only?

**Reply by wereHamster** [reply]
@tettoffensive In the webpack config you can add code to the entry points conditional on `isServer`, eg.

```
  webpack: (config, options) => {
    const originalEntry = config.entry
    config.entry = async () => {
      const entries = await originalEntry()

      if (entries['main.js']) {
        // Load polyfills only in the client.
        if (!options.isServer && !entries['main.js'].includes('./src/polyfills.ts')) {
          entries['main.js'].unshift('./src/polyfills.ts')
        }
      }

      return entries
    }

    return config
  }
```

**Reply by charkour** [reply]
Is there a way to only include a polyfill if the browser does not support the feature? We are trying to shave bytes on our bundle size. Also, Google Lighthouse audits suggest we only ship modern JS to the client and our team is guessing the old JS shipped is due to the polyfill. Thanks.

**Reply by disjukr** [reply]
> > Also note that if you import polyfills in _app, and have have a code editor plugin which automatically sorts imports, you may end up in a situation that the polyfill is loaded too late, after code that needs it. Therefore I'm not a fan of this solution.
> 
> I disagree. A single line import is much easier than overriding the webpack config. You can just add a lint ignore to that line?

@leerob Some mobile browsers that our customers using are throwing errors due to the absence of `globalThis`, which Next.js runtime relies on. To polyfill this, we need to ensure that our polyfill runs before Next.js scripts.

Resolving this issue is not possible with `import` statements, and even when trying to place `<Script>` or `<script>` at the top of `layout.js`, Next.js runtime always loads first, making it impossible to address this problem properly.

Is there any recommended way to solve this?

**Comment by andresgutgon** [comment]
Hi, I'm having this issue but with Next 13.x and using the `app` directory. What's the recommended way of using pollyfills with new `app` folder? 

Thanks!

**Reply by MartinXPN** [reply]
I've created a client component and added the polyfills in a `useEffect` to ensure that this part executes on a client. Otherwise, the components are server-only by default. So, the javascript doesn't get shipped to the client.
It would've been great to see this in docs.

**Reply by MartinXPN** [reply]
I actually tried just adding `'use client';` at the top of the `polyfills.ts` file and importing it in the root `layout.tsx` and it worked fine. 
Seems like there is no need of a separate client component and a `useEffect`.

**Reply by andresgutgon** [reply]
Nice!

**Reply by AjaxTheOne** [reply]
Thanks, it solved my issue with "reflect-metadata"!

**Reply by TracerLee** [reply]
@MartinXPN I found that I cannot import core-js related polyfills using this approach, which would result in an error message like "Module parse failed: Cannot use 'import.meta' outside a module".
Example:
`polyfills.ts`
```ts
'use client';
import 'core-js/modules/es.array.every'
```

**Reply by plrthink** [reply]
I also found this doesn't work since the code is not being bundled in client scripts.

**Reply by carlos-dubon** [reply]
@leerob where should we `import 'reflect-metadata';` for reflect polyfill to work?

**Reply by secure12** [reply]
@MartinXPN `"use client";` is the magic! Thank you!

**Comment by RaenonX** [comment]
https://github.com/RaenonX-PokemonSleep/pokemon-sleep-ui/commit/d4ce3c1862ed5cef7ddcb1e83b37ece66ab0300a

Using next v14. Tried to add custom polyfill like what was shown in this thread in the commit above, but it doesn't seem like it's getting polyfilled. How do I check if it's really getting polyfilled?

Edit: Solved. Need these changes and all good. https://github.com/RaenonX-PokemonSleep/pokemon-sleep-ui/commit/b9fb52339a085f7ed632688dd5fb9278e8f862eb

**Reply by OptimisticQuan** [reply]
Thank you, this solved a problem that had been bothering me for half a day.

**Comment by ArnabChatterjee20k** [comment]
Any docs for importing custom webpack polyfills in next js 14?? I am trying to solve the issue for node:buffer error

**Reply by fireMan-34** [reply]
Importing static polyfills using Head in public templates or layouts may be a solution.

**Reply by ArnabChatterjee20k** [reply]
Any code example or repo example?

**Reply by RaenonX** [reply]
![image](https://github.com/vercel/next.js/assets/18443743/edc37170-8063-4560-83d0-761e60eb1241)

Sorry had to screenshot because my repo is private now...lol

**Reply by ArnabChatterjee20k** [reply]
No isssue. I will figure it out from the image
Thanks

**Reply by ArnabChatterjee20k** [reply]
Basically you have imported the polyfills in the root or the top level layout

**Reply by fireMan-34** [reply]
I tested it and found that exporting with the Head component only works in page routing mode.

this is `pages/_document.tsx`
![image](https://github.com/vercel/next.js/assets/79759571/efe21939-866e-46d1-9c4d-b008ed410a36)
this is `app/_layout.tsx`
![image](https://github.com/vercel/next.js/assets/79759571/57f27642-9c6f-4bc7-87b1-74443b820ea5)

this is build output
this is page router output
![image](https://github.com/vercel/next.js/assets/79759571/18ea68e9-ca3a-4723-8c5b-67d84c927b2a)

this is app router output
![image](https://github.com/vercel/next.js/assets/79759571/eecf88e7-8bae-4221-a945-a033eaaa2905)

**Comment by osdiab** [comment]
As the documentation is out of date and a clear answer to how you're supposed to do this in modern NextJS versions is also not clear, I've opened a documentation issue regarding custom polyfills.

https://github.com/vercel/next.js/issues/74730

**Comment by popuguytheparrot** [comment]
how to disable polyfill to do not create polyfills chunk?
